### PR TITLE
[Optimize] Support adaptive runtime filter(#7546)

### LINKED ISSUE #7546: [Enhancement] Support merge IN predicate & Adaptive runtime filter
### Search before asking

- [X] I had searched in the [issues](https://github.com/apache/incubator-doris/issues?q=is%3Aissue) and found no similar issues.

## 问题

1. 当前的runtime filter在shuffle join时不支持合并IN predicate。
2. IN predicate存在一个比较奇怪的逻辑：如果遇到remote target则会隐式地把IN改成Bloom filter。这和用户设定的`runtime_filter_type='IN'`不一致，会导致用户迷惑。
3. IN predicate和其他filter通常不会一起用，因为IN predicate本身就是一个精确的过滤条件，已经足以把数据提前过滤好，如果加上其他filter，那么其他filter可能会做无用功。但现在没法在运行时自适应的决定使用IN predicate还是其他filter。

## 目标
1.  支持IN predicate的合并。
2. 去掉IN predicate隐式转变成Bloom filter的逻辑。
3.  支持一种自适应filter，在运行时选择应该使用哪种filter。

注：由于不容易判断什么情况适合使用MIN MAX filter，因此自适应filter仅在运行时判断使用IN predicate还是Bloom filter，判断依据是in过滤值数量少于runtime_filter_max_in_num时，尽量用IN predicate而不是Bloom filter，否则使用Bloom filter。

### Solution

_No response_

### Are you willing to submit PR?

- [X] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

### PR DESCRIPTION
**Pull request by 924060929**
## Proposed changes

Close related #7546.

变更1：支持一种自适应的runtime filter：IN_OR_BLOOM_FILTER
处理逻辑为
1. 如果右表的行数 < runtime_filter_max_in_num，则IN predicate能生效
2. 如果右表的行数 >= runtime_filter_max_in_num，则Bloom filter能生效

变更2：默认的runtime filter修改为filter：IN_OR_BLOOM_FILTER

## Types of changes

What types of changes does your code introduce to Doris?
_Put an `x` in the boxes that apply_

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] Documentation Update (if none of the other choices apply)
- [ ] Code refactor (Modify the code structure, format the code, etc...)
- [x] Optimization. Including functional usability improvements and performance improvements.
- [ ] Dependency. Such as changes related to third-party components.
- [ ] Other.

## Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code._

- [x] I have created an issue on (Fix #7546) and described the bug/feature there in detail
- [x] Compiling and unit tests pass locally with my changes
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] If these changes need document changes, I have updated the document
- [x] Any dependent changes have been merged

## Further comments

If this is a relatively large or complex change, kick off the discussion at dev@doris.apache.org by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

File: fe/fe-core/src/test/java/org/apache/doris/planner/QueryPlanTest.java [hunk 1]
```diff
@@ -1366,23 +1366,15 @@ public void testRuntimeFilterMode() throws Exception {
 
         queryStr = "explain select * from jointest t2, jointest t1 where t1.k1 <=> t2.k1";
         Deencapsulation.setField(connectContext.getSessionVariable(), "runtimeFilterMode", "LOCAL");
-        Deencapsulation.setField(connectContext.getSessionVariable(), "runtimeFilterType", 7);
+        Deencapsulation.setField(connectContext.getSessionVariable(), "runtimeFilterType", 15);
```
Comment (EmmyMiao87) [inline fe/fe-core/src/test/java/org/apache/doris/planner/QueryPlanTest.java:1369]: What is the main purpose of this single test?
Type=15 is meaningless, writing this way will give users a wrong guide.
Can I change the type? Such as 8

File: fe/fe-core/src/test/java/org/apache/doris/planner/QueryPlanTest.java [hunk 2]
```diff
@@ -1429,6 +1421,64 @@ public void testRuntimeFilterType() throws Exception {
         explainString = UtFrameUtils.getSQLPlanOrErrorMsg(connectContext, queryStr);
         Assert.assertTrue(explainString.contains("runtime filters: RF000[in] <- `t1`.`k1`, RF001[bloom] <- `t1`.`k1`, RF002[min_max] <- `t1`.`k1`"));
         Assert.assertTrue(explainString.contains("runtime filters: RF000[in] -> `t2`.`k1`, RF001[bloom] -> `t2`.`k1`, RF002[min_max] -> `t2`.`k1`"));
+
+        Deencapsulation.setField(connectContext.getSessionVariable(), "runtimeFilterType", 8);
+        explainString = UtFrameUtils.getSQLPlanOrErrorMsg(connectContext, queryStr);
+        Assert.assertTrue(explainString.contains("runtime filters: RF000[in_or_bloom] <- `t1`.`k1`"));
+        Assert.assertTrue(explainString.contains("runtime filters: RF000[in_or_bloom] -> `t2`.`k1`"));
+
+        Deencapsulation.setField(connectContext.getSessionVariable(), "runtimeFilterType", 9);
```
Comment (EmmyMiao87) [inline fe/fe-core/src/test/java/org/apache/doris/planner/QueryPlanTest.java:1430]: The type value 9 is not used in the scene, you can make a note in the front. This case is only used to test the logical correctness. In fact, the value of type is meaningless.

File: fe/fe-core/src/test/java/org/apache/doris/qe/RuntimeFilterTypeHelperTest.java [hunk 3]
```diff
@@ -50,6 +50,30 @@ public void testNormal() throws DdlException {
         runtimeFilterType = "IN,BLOOM_FILTER,MIN_MAX";
         Assert.assertEquals(new Long(7L), RuntimeFilterTypeHelper.encode(runtimeFilterType));
 
+        runtimeFilterType = "IN_OR_BLOOM_FILTER";
+        Assert.assertEquals(new Long(8L), RuntimeFilterTypeHelper.encode(runtimeFilterType));
+
+        runtimeFilterType = "IN,IN_OR_BLOOM_FILTER";
+        Assert.assertEquals(new Long(9L), RuntimeFilterTypeHelper.encode(runtimeFilterType));
```
Comment (EmmyMiao87) [inline fe/fe-core/src/test/java/org/apache/doris/qe/RuntimeFilterTypeHelperTest.java:57]: same as above

File: docs/zh-CN/administrator-guide/runtime-filter.md [hunk 4]
```diff
@@ -138,6 +138,9 @@ set runtime_filter_type=7;
     - 默认只有右表数据行数少于1024才会下推（可通过session变量中的`runtime_filter_max_in_num`调整）。
     - 目前IN predicate已实现合并方法。
     - 当同时指定In predicate和其他filter，并且in的过滤数值没达到runtime_filter_max_in_num时，会尝试把其他filter去除掉。原因是In predicate是精确的过滤条件，即使没有其他filter也可以高效过滤，如果同时使用则其他filter会做无用功。目前仅在Runtime filter的生产者和消费者处于同一个fragment时才会有去除非in filter的逻辑。
+- **IN or Bloom Filter**: 根据右表的值的行数来判断使用IN predicate还是Bloom Filter
+    - 默认在右表数据行数少于1024时会使用IN predicate（可通过session变量中的`runtime_filter_max_in_num`调整）
+    - 否则使用Bloom filter
 - **
```
Comment (EmmyMiao87) [inline docs/zh-CN/administrator-guide/runtime-filter.md:145]: 是否可以给个最佳实践，引导用户使用。
1. IN/BF, 8
2. min max, 4

其他类型的runtime filter type，属于『高级用法』目录下。

File: fe/fe-core/src/test/java/org/apache/doris/planner/QueryPlanTest.java (hunk 1 above)
Comment (924060929) [reply fe/fe-core/src/test/java/org/apache/doris/planner/QueryPlanTest.java:1369]: > Can I change the type? Such as 8

You can.
origin test, 7 means: IN,BLOOM_FILTER,MIN_MAX
new test, 15 means: IN,BLOOM_FILTER,MIN_MAX,IN_OR_BLOOM_FILTER

Already test `8` in other test

File: fe/fe-core/src/test/java/org/apache/doris/planner/QueryPlanTest.java (hunk 2 above)
Comment (924060929) [reply fe/fe-core/src/test/java/org/apache/doris/planner/QueryPlanTest.java:1430]: Okey, I will add some note to the doc

Comment (924060929) [reply fe/fe-core/src/test/java/org/apache/doris/planner/QueryPlanTest.java:1430]: add add the comment to this line

File: be/src/exprs/runtime_filter.cpp [hunk 5]
```diff
@@ -351,6 +359,26 @@ class RuntimePredicateWrapper {
...
+        if (_hybrid_set->size() > 0) {
+            auto it = _hybrid_set->begin();
+            while (it->has_next()) {
+                _bloomfilter_func->insert(it->get_value());
+                it->next();
+            }
+            // release in filter
+            _hybrid_set.reset(create_set(_column_return_type));
+        }
+        return Status::OK();
```
Comment (EmmyMiao87) [inline be/src/exprs/runtime_filter.cpp:379]: Please update the runtime filter type in profile ~

File: be/src/exprs/runtime_filter_slots.h [hunk 6]
```diff
@@ -60,7 +60,27 @@ class RuntimeFilterSlotsBase {
...
+                return 1;
+            } else if (d1.type == TRuntimeFilterType::IN_OR_BLOOM) {
+                return -1;
+            } else if (d2.type == TRuntimeFilterType::IN_OR_BLOOM) {
+                return 1;
+            } else {
+                return d1.type < d2.type ? -1 : 1;
+            }
+        };
+        std::sort(sorted_runtime_filter_descs.begin(), sorted_runtime_filter_descs.end(), compare_desc);
```
Comment (BiteTheDDDDt) [inline be/src/exprs/runtime_filter_slots.h:81]: It is ok to `compare_desc` return `-1` or `1`?
https://en.cppreference.com/w/cpp/named_req/Compare

**Comment by BiteTheDDDDt** [comment]
Is it possible to construct two bloom filter when open `BLOOM_FILTER` and `IN_OR_BLOOM_FILTER` ?

**Comment by 924060929** [comment]
> Is it possible to construct two bloom filter when open `BLOOM_FILTER` and `IN_OR_BLOOM_FILTER` ?

It could be. Should we remove the second bloom filter in local mode？In current implement，we only ignore filter in local target. When exist remote target, merge node can not ignore other filter because of different filter in a join node send different merge request. The merge bloom filter could happens before merge in filter, so the in filter and bloom filter could be  both applied.

**Comment by EmmyMiao87** [comment]
Or can we be a little more aggressive and keep only IN/BF and min max types.
In the documentation, users are only made aware of these two types.
Four types of IN, BF, MIN_MAX, IN/BF are reserved in the code, only for developer code debugging?

**Comment by 924060929** [comment]
> 

okey

**Comment by 924060929** [comment]
> Or can we be a little more aggressive and keep only IN/BF and min max types. In the documentation, users are only made aware of these two types. Four types of IN, BF, MIN_MAX, IN/BF are reserved in the code, only for developer code debugging?

okey

File: be/src/exprs/runtime_filter.cpp [hunk 7]
```diff
@@ -351,6 +359,26 @@ class RuntimePredicateWrapper {
         return Status::OK();
     }
 
+    Status change_to_bloom_filter() {
+        if (_filter_type != RuntimeFilterType::IN_OR_BLOOM_FILTER) {
```
Comment (morningman) [inline be/src/exprs/runtime_filter.cpp:363]: Looks like there is no where to check this Status?
And better not using `stringstream` in hot path.

File: be/src/exprs/runtime_filter.cpp [hunk 8]
```diff
@@ -439,9 +481,20 @@ class RuntimePredicateWrapper {
...
+        bool can_not_merge_in_or_bloom
+            = _filter_type == RuntimeFilterType::IN_OR_BLOOM_FILTER &&
+                  (wrapper->_filter_type != RuntimeFilterType::IN_FILTER
+                   && wrapper->_filter_type != RuntimeFilterType::BLOOM_FILTER);
+
+        bool can_not_merge_other = _filter_type != RuntimeFilterType::IN_OR_BLOOM_FILTER
+                               && _filter_type != wrapper->_filter_type;
+
+        if (can_not_merge_in_or_bloom || can_not_merge_other) {
+            std::stringstream msg;
```
Comment (morningman) [inline be/src/exprs/runtime_filter.cpp:493]: Do not using `stringstream` here.
Same as others

File: be/src/exprs/runtime_filter.cpp [hunk 9]
```diff
@@ -483,6 +536,63 @@ class RuntimePredicateWrapper {
...
+                        LOG(INFO) << msg.str();
+                        change_to_bloom_filter();
+                    }
+                // in merge bloom filter
+                } else {
+                    std::stringstream msg;
+                    msg << "fragment instance " << _fragment_instance_id.to_string()
+                        << " change runtime filter to bloom filter(id=" << _filter_id
+                        << ") because: already exist a bloom filter";
+                    LOG(INFO) << msg.str();
```
Comment (morningman) [inline be/src/exprs/runtime_filter.cpp:569]: Do not print log here

File: be/src/exprs/runtime_filter.cpp [hunk 10]
```diff
@@ -483,6 +536,63 @@ class RuntimePredicateWrapper {
...
+                        return Status::InvalidArgument(msg.str());
+                    }
+                    _hybrid_set->insert(wrapper->_hybrid_set.get());
+                    if (_max_in_num >= 0 && _hybrid_set->size() >= _max_in_num) {
+                        std::stringstream msg;
+                        msg << "fragment instance " << _fragment_instance_id.to_string()
+                            << " change runtime filter to bloom filter(id=" << _filter_id
+                            << ") because: in_num(" << _hybrid_set->size()
+                            << ") >= max_in_num(" << _max_in_num << ")";
+                        LOG(INFO) << msg.str();
```
Comment (morningman) [inline be/src/exprs/runtime_filter.cpp:560]: Do not print log here

File: be/src/exprs/runtime_filter.cpp (hunk 9 above)
Comment (924060929) [reply be/src/exprs/runtime_filter.cpp]: already change to VLOG_DEBUG

File: be/src/exprs/runtime_filter.cpp (hunk 10 above)
Comment (924060929) [reply be/src/exprs/runtime_filter.cpp]: already change to VLOG_DEBUG

File: be/src/exprs/runtime_filter.cpp (hunk 8 above)
Comment (924060929) [reply be/src/exprs/runtime_filter.cpp]: already change to CHECK

File: be/src/exprs/runtime_filter.cpp (hunk 7 above)
Comment (924060929) [reply be/src/exprs/runtime_filter.cpp]: already change to return void and check by CHECK

File: be/src/exprs/runtime_filter_slots.h (hunk 6 above)
Comment (924060929) [reply be/src/exprs/runtime_filter_slots.h:81]: done

File: be/src/exprs/runtime_filter.cpp (hunk 5 above)
Comment (924060929) [reply be/src/exprs/runtime_filter.cpp]: done

File: docs/zh-CN/administrator-guide/runtime-filter.md (hunk 4 above)
Comment (924060929) [reply docs/zh-CN/administrator-guide/runtime-filter.md:143]: 后续再加最佳实践

File: docs/en/administrator-guide/runtime-filter.md [hunk 11]
```diff
@@ -138,6 +138,9 @@ set runtime_filter_type=7;
     - By default, only the number of data rows in the right table is less than 1024 will be pushed down (can be adjusted by `runtime_filter_max_in_num` in the session variable).
     - Currently IN predicate already implement a merge method.
     - When IN predicate and other filters are specified at the same time, and the filtering value of IN predicate does not reach runtime_filter_max_in_num will try to remove other filters. The reason is that IN predicate is an accurate filtering condition. Even if there is no other filter, it can filter efficiently. If it is used at the same time, other filters will do useless work. Currently, only when the producer and consumer of the runtime filter are in the same fragment can there be logic to remove the Non-IN predicate.
+- **IN_OR_BLOOM FILTER**:
```
[the remark this answers is not in the document]
Comment (924060929) [reply docs/en/administrator-guide/runtime-filter.md]: done

File: docs/zh-CN/administrator-guide/runtime-filter.md [hunk 12]
```diff
@@ -138,6 +138,9 @@ set runtime_filter_type=7;
     - 默认只有右表数据行数少于1024才会下推（可通过session变量中的`runtime_filter_max_in_num`调整）。
     - 目前IN predicate已实现合并方法。
     - 当同时指定In predicate和其他filter，并且in的过滤数值没达到runtime_filter_max_in_num时，会尝试把其他filter去除掉。原因是In predicate是精确的过滤条件，即使没有其他filter也可以高效过滤，如果同时使用则其他filter会做无用功。目前仅在Runtime filter的生产者和消费者处于同一个fragment时才会有去除非in filter的逻辑。
+- **IN or Bloom Filter**: 根据右表的值的行数来判断使用IN predicate还是Bloom Filter
```
[the remark this answers is not in the document]
Comment (924060929) [reply docs/zh-CN/administrator-guide/runtime-filter.md]: done

[the remark this answers is not in the document]
Comment (924060929) [reply docs/zh-CN/administrator-guide/runtime-filter.md]: done

File: fe/fe-core/src/test/java/org/apache/doris/qe/RuntimeFilterTypeHelperTest.java (hunk 3 above)
Comment (924060929) [reply fe/fe-core/src/test/java/org/apache/doris/qe/RuntimeFilterTypeHelperTest.java:57]: Just for test, we should guide user by doc.

**Review by EmmyMiao87 — APPROVED** [review]
LGTM

**Pull request merged**